lambda_functions/two_way_sms/two_way_sms_v2.py

```diff
@@ -0,0 +1,344 @@
+#            'self_managed': <value>,
+#        }
+#    }
+################################################################################################
+
+if os.getenv("NOTIFY_ENVIRONMENT") == "test":
+    two_way_sms_table_dict = {}
+else:
+    try:
+        logger.info("Retrieving the 10DLC-to-URL mapping from the database . . .")
+        db_connection = psycopg2.connect(sqlalchemy_database_uri)
+        if db_connection is None:
+            sys.exit("No database connection.")
+
+        with db_connection.cursor() as c:
+            # https://www.psycopg.org/docs/cursor.html#cursor.execute
+            c.execute(INBOUND_NUMBERS_QUERY)
+            mapping_data = c.fetchall()
+            logger.debug("Data returned from query: %s", mapping_data)
+
+        db_connection.close()
+        logger.info(". . . Retrieved the 10DLC-to-URL mapping from the database.")
+    except psycopg2.Warning as e:
+        logger.warning(e)
+    except psycopg2.OperationalError as e:        
+        logger.exception(e)
+        sys.exit("Unable to retrieve the 10DLC-to-URL mapping from the database.")
+
+    # Create the mapping table with a generator expression.
+    two_way_sms_table_dict = {
+        n: {
+            'service_id': s,
+            'url_endpoint': u,
+            'self_managed': False if sm == None else sm,
+        } for n, s, u, sm in mapping_data
+    }
+
+logger.debug('Two way table as a dictionary with numbers as keys: %s', two_way_sms_table_dict)
+
+################################################################################################
+
+#aws_pinpoint_client = boto3.client('pinpoint', region_name=AWS_REGION)
+aws_sqs_client = boto3.client('sqs', region_name=AWS_REGION)
+
+
+# ------------------------------------------- Begin Invocation --------------------------------------------
+def notify_incoming_sms_handler(event: dict, context: any):
+    """
+    Handler for inbound messages from SQS.
+    """
+
+    batch_item_failures = []
+
+    if not valid_event(event):
+        logger.critical("Invalid event: %s", event)
+
+        # Push the message to the dead letter queue, and return 200 to have the message removed from feeder queue.
+        push_to_sqs(event, False)
+        return create_response(200)
+
+    # SQS events should contain "Records".  This is checked in the call above to valid_event.
+    for record in event["Records"]:
+        logger.info("Processing an SQS inbound_sms record . . .")
+
+        # SQS event records should contain "body".  This is checked in the call above to valid_event.
+        record_body = record["body"]
+
+        try:
+            record_body = json.loads(record_body)
+            inbound_sms = json.loads(record_body.get("Message", ''))
+        except (json.decoder.JSONDecodeError, TypeError) as e:
+            # Deadletter
+            push_to_sqs(record_body, False)
+            logger.error("Malformed record body.")
+            logger.exception(e)
```

Comment:
Does this need the `("%s", e)`?

Reply:
Yes it does.

Reply:
It is not necessary.  With "exception", I just want to output the stack trace.
```
>>> import logging
>>> logging.error(ValueError("Log this exception."))
ERROR:root:Log this exception.
```